File: dsa_leetcode_beginner/0139-word-break/0139-word-break.java
import java.util.*;

class Solution {
    public boolean wordBreak(String s, List<String> wordDict) {
        Set<String> wordSet = new HashSet<>(wordDict);
        boolean[] res = new boolean[s.length() + 1];
        res[0] = true;

        for (int i = 0; i <= s.length(); i++) {
            for (String word : wordSet) {
                if (res[i] && i + word.length() <= s.length() && s.substring(i, i + word.length()).equals(word)) {
                    res[i + word.length()] = true;
                }
            }
        }

        return res[s.length()];
    }
}

File: dsa_leetcode_beginner/0139-word-break/0139-word-break.js
var wordBreak = function(s, wordDict) {
    const wordSet = new Set(wordDict);
    const res = Array(s.length + 1).fill(false);
    res[0] = true;

    for (let i = 0; i <= s.length; i++) {
        for (const word of wordSet) {
            if (res[i] && i + word.length <= s.length && s.substring(i, i + word.length) === word) {
                res[i + word.length] = true;
            }
        }
    }

    return res[s.length];
};
